src/forces.jl

```diff
@@ -0,0 +1,74 @@
+"""
+    forces = forcecomponents(seq::Vector{PDBResidue}, keep::AbstractVector{Bool}; kwargs...)
+
+Calculate the forces between residues in a `PDBResidue` sequence `seq` based on
+the features of each residue. `kwargs` are passed to
+`add_features_from_residue!`.
+
+Upon return, there is one force-matrix for each residue in `seq` for which
+`keep` is `true`. Each force-matrix is a 3x5 matrix where each row corresponds
+to a force component (x, y, z) and each column corresponds to a feature type
+(steric, hydrophobic, ionic, hbond, aromatic).
+"""
+function forcecomponents(seq::Vector{PDBResidue}, keep::AbstractVector{Bool}; kwargs...)
+    # Create a MGMM for each residue
+    mgmms = [IsotropicMultiGMM(Dict{Symbol,IsotropicGMM{3,Float64}}()) for _ in seq]
+    for i in eachindex(seq)
+        add_features_from_residue!(mgmms[i], seq[i]; kwargs...)
+    end
+    # steric = 1, hydrophobic = 2, ionic = 3, hbond = 4, aromatic = 5
+    forces = Matrix{Float64}[]
+    for i in eachindex(seq)
+        keep[i] || continue
+        force = zeros(3, 5)
+        mgmmi = mgmms[i]
+        for (colidx, keyi) in ((1, :Steric), (2, :Hydrophobe), (3, :PosIonizable), (3, :NegIonizable), (4, :Donor), (4, :Acceptor), (5, :Aromatic))
+            haskey(mgmmi.gmms, keyi) || continue
+            gx = mgmmi.gmms[keyi]
+            for j in eachindex(seq)
+                if i == j
+                    continue
+                end
+                mgmmj = mgmms[j]
+                if keyi ∈ (:Steric, :Hydrophobe, :Aromatic)
+                    keyj = keyi
+                    haskey(mgmmj.gmms, keyj) || continue
+                    force!(@view(force[:, colidx]), gx, mgmmj.gmms[keyj]; coef = keyi === :Steric ? -1 : 1)
+                elseif keyi ∈ (:PosIonizable, :NegIonizable)
+                    for keyj in (:PosIonizable, :NegIonizable)
+                        haskey(mgmmj.gmms, keyj) || continue
+                        force!(@view(force[:, colidx]), gx, mgmmj.gmms[keyj]; coef = (keyj == keyi ? -1 : 1))
+                    end
```

Comment:
Is it worth considering any interactions between aromatic and cationic features here?

Reply:
I would love to do that (we'd have to add support in `overlap` too). Currently I am unsure how to handle the aromatic interactions. Aromatic-aromatic can be either attractive or repulsive depending on the displacement and orientation of the aromatic rings. In my current usage, I'm setting that coefficient to 0, i.e., ignoring aromatic interactions.

But cation-pi might actually be easier, as that should (I think) be purely attractive.